Sources/SemanticIndex/UpdateIndexStoreTaskDescription.swift

```diff
@@ -0,0 +1,311 @@
+    self.buildSystemManager = buildSystemManager
+    self.index = index
+    self.didFinishCallback = didFinishCallback
+  }
+
+  public func execute() async {
+    defer {
+      didFinishCallback(self)
+    }
+    // Only use the last two digits of the indexing ID for the logging scope to avoid creating too many scopes.
+    // See comment in `withLoggingScope`.
+    // The last 2 digits should be sufficient to differentiate between multiple concurrently running indexing operation.
+    await withLoggingSubsystemAndScope(
+      subsystem: "org.swift.sourcekit-lsp.indexing",
+      scope: "update-indexstore-\(id % 100)"
+    ) {
+      let startDate = Date()
+
+      let filesToIndexDescription = filesToIndex.map { $0.fileURL?.lastPathComponent ?? $0.stringValue }
+        .joined(separator: ", ")
+      logger.log(
+        "Starting updating index store with priority \(Task.currentPriority.rawValue, privacy: .public): \(filesToIndexDescription)"
+      )
+      let filesToIndex = filesToIndex.sorted(by: { $0.stringValue < $1.stringValue })
+      // TODO (indexing): Once swiftc supports it, we should group files by target and index files within the same
+      // target together in one swiftc invocation.
+      for file in filesToIndex {
+        await updateIndexStoreForSingleFile(file)
+      }
+      logger.log(
+        "Finished updating index store in \(Date().timeIntervalSince(startDate) * 1000, privacy: .public)ms: \(filesToIndexDescription)"
+      )
+    }
+  }
+
+  public func dependencies(
+    to currentlyExecutingTasks: [UpdateIndexStoreTaskDescription]
+  ) -> [TaskDependencyAction<UpdateIndexStoreTaskDescription>] {
+    return currentlyExecutingTasks.compactMap { (other) -> TaskDependencyAction<UpdateIndexStoreTaskDescription>? in
+      guard !other.filesToIndex.intersection(filesToIndex).isEmpty else {
+        // Disjoint sets of files can be indexed concurrently.
+        return nil
+      }
+      if self.filesToIndex.count < other.filesToIndex.count {
+        // If there is an index operation with more files already running, suspend it.
+        // The most common use case for this is if we schedule an entire target to be indexed in the background and then
+        // need a single file indexed for use interaction. We should suspend the target-wide indexing and just index
+        // the current file to get index data for it ASAP.
+        return .cancelAndRescheduleDependency(other)
+      } else {
+        return .waitAndElevatePriorityOfDependency(other)
+      }
+    }
+  }
+
+  private func updateIndexStoreForSingleFile(_ uri: DocumentURI) async {
+    guard let url = uri.fileURL else {
+      // The URI is not a file, so there's nothing we can index.
+      return
+    }
+    guard !index.hasUpToDateUnit(for: url) else {
+      // We consider a file's index up-to-date if we have any up-to-date unit. Changing build settings does not
+      // invalidate the up-to-date status of the index.
```

Comment:
I guess you have to `touch` the file or something if you want to index it with up-to-date build settings?

Reply:
Still an open question: https://github.com/apple/sourcekit-lsp/issues/1251